File: ScoreBoostPowerUp.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */
package com.mycompany.mazegame;

/**
 *
 * @author lsuti
 */
import java.awt.image.BufferedImage;

public class ScoreBoostPowerUp extends PowerUp {

    private static final int MAX_MULTIPLIER = 5;
    private static final int DURATION = 100;
    private static boolean multiplierActive = false;
    private static int currentMultiplier = 1;
    private static int multiplierTimeRemaining = 0;

    public ScoreBoostPowerUp(BufferedImage img, int x, int y) {
        super(img, x, y);
    }

    @Override
    public void update(GamePanel panel, boolean collisionEnabled) {
        if (multiplierActive) {
            multiplierTimeRemaining--;
            if (multiplierTimeRemaining <= 0) {
                currentMultiplier = 1;
                multiplierActive = false;
            }
        }
    }

    @Override
    public void applyEffect(Player player, GamePanel panel) {
        // Reset duration if already active
        if (multiplierActive) {
            multiplierTimeRemaining = DURATION;
        }

        // Calculate bonus using the player's combo multiplier
        int bonusScore = player.getComboMultiplier()
                * panel.getScoreBoostBase()
                * currentMultiplier;
        player.addScore(bonusScore);

        // Increase our internal multiplier up to the max
        if (currentMultiplier < MAX_MULTIPLIER) {
            currentMultiplier++;
        }
        multiplierActive = true;
        multiplierTimeRemaining = DURATION;

        // Add a time bonus based on the score boost
        int timeBonus = bonusScore / panel.getTimeScoreRatio();
        panel.setTimeRemaining(panel.getTimeRemaining() + timeBonus);

        // Advance the player's combo and record the power‐up use
        player.incrementComboMultiplier();
        player.getStatsTracker().incrementPowerUpsUsed();
    }
}

File: ScoreBoostPowerUp.ts
import { PowerUp } from './PowerUp';
import { Player } from './Player';
import { GamePanel } from './GamePanel';

export class ScoreBoostPowerUp extends PowerUp {
    private static readonly MAX_MULTIPLIER = 5;
    private static readonly DURATION = 100;
    private static multiplierActive = false;
    private static currentMultiplier = 1;
    private static multiplierTimeRemaining = 0;

    update(panel: GamePanel, collisionEnabled: boolean): void {
        if (ScoreBoostPowerUp.multiplierActive) {
            ScoreBoostPowerUp.multiplierTimeRemaining--;
            if (ScoreBoostPowerUp.multiplierTimeRemaining <= 0) {
                ScoreBoostPowerUp.currentMultiplier = 1;
                ScoreBoostPowerUp.multiplierActive = false;
            }
        }
    }

    applyEffect(player: Player, panel: GamePanel): void {
        // Activate and reset the multiplier timer
        ScoreBoostPowerUp.multiplierActive = true;
        ScoreBoostPowerUp.multiplierTimeRemaining = ScoreBoostPowerUp.DURATION;

        // Compute bonus: combo * base score * current multiplier
        const bonus =
            player.combo *
            panel.scoreBoostBase *
            ScoreBoostPowerUp.currentMultiplier;
        player.addScore(bonus);

        // Bump multiplier for next use (capped)
        if (ScoreBoostPowerUp.currentMultiplier < ScoreBoostPowerUp.MAX_MULTIPLIER) {
            ScoreBoostPowerUp.currentMultiplier++;
        }

        // Convert part of that bonus into extra time
        const timeBonus = bonus / panel.timeScoreRatio;
        panel.timeRemaining += timeBonus;

        // Update player state
        player.incrementComboMultiplier();
        player.stats.incrementPowerUpsUsed();
    }
}